libs/wingsdk/src/target-sim/resource.ts

```diff
@@ -1,5 +1,308 @@
+
+    if (process.env.WING_TARGET) {
+      this.addEnvironment("WING_TARGET", process.env.WING_TARGET);
+    }
+
+    this.factory = factory;
+  }
+
+  public addPermission(resource: IStdResource, op: string): void {
+    this.permissions.push([resource, op]);
+  }
+
+  /** @internal */
+  public get _liftMap(): LiftMap {
+    return {
+      [ResourceInflightMethods.CALL]: [],
+    };
+  }
+
+  /**
+   * Obtain a token that can be used to reference an attribute of this
+   * resource that is only resolved once the resource is started in the simulator.
+   *
+   * If the token is used in inflight code or in the configuration of a simulated
+   * resource (e.g. as an environment variable), the relevant resource will
+   * automatically take a dependency on the resource the attribute belongs to.
+   *
+   * @param name The name of the attribute.
+   * @returns An attribute token.
+   */
+  public attrToken(name: string): string {
+    return simulatorAttrToken(this, name);
+  }
+
+  /** @internal */
+  public _preSynthesize(): void {
+    super._preSynthesize();
+
+    const inflightClient = this.factory._toInflight();
+    const code = `\
+        "use strict";
+        let $klass;
+        exports.start = async function(statedir) {
+          if ($klass) {
+            throw Error('resource already started');
+          }
+          const attrs = {};
+          const ctx = {};
+          ctx.statedir = () => statedir;
+          ctx.resolveAttr = (name, value) => attrs[name] = value;
+          ctx.log = (message, level) => {
+            if (!level) level = 'info';
+            console.log(level + ': ' + message);
+          };
+          const client = ${inflightClient};
+          const noop = () => {};
+          const klass = (await client.handle()) ?? noop;
+          await klass.onStart(ctx);
+          ctx.resolveAttr = () => {
+            throw Error('cannot resolve attributes outside of onStop method');
+          };
+          $klass = klass;
+          return attrs;
+        };
+
+        exports.call = async function(propName, ...args) {
+          if (!$klass) {
+            throw Error('resource not started');
+          }
+          if (propName === 'onStart') {
+            throw Error('Cannot call "onStart"');
+          }
+          if (propName === 'onStop') {
+            throw Error('Cannot call "onStop"');
+          }
+          const prop = $klass[propName];
+          if (!prop) {
+            throw Error('Method or property "' + propName + '" not found');
+          }
+          if (typeof prop !== 'function') {
+            if (args.length > 0) {
+              throw Error('Property "' + propName + '" is not a function');
+            }
+            return prop;
+          }
+          return await prop.call($klass, ...args);
+        };
+
+        exports.stop = async function() {
+          if (!$klass) {
+            throw Error('resource not started');
+          }
+          await $klass.onStop();
```

Comment:
Can we put `onStart` and `onStop` in constants so it's not that fragile?

Also - is it possible to put most of this code outside in some helper function that's properly compiled and unit-tested?

Reply:
I'm not sure how to do this without adding more layers of runtime indirection. If you have ideas though, feel free to push something to the PR, it sounds like a nice improvement direction.